### PR TITLE
Do not mutate `SymNode` expression.

### PR DESCRIPTION
**Pull request by ysiraichi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #107493
* __->__ #107492
* #107523

This PR stops `SymNode` from mutating (i.e. simplifying) its expression. Instead, the
simplification (without mutation) is deferred to the `SymNode.maybe_as_int` method.

```python
- FakeTensor(size=(s0,), ...)
- FakeTensor(size=(s1, s2, s3), ...)

- Eq(s0, s1 + s2 + s3)

- FakeTensor(size=(s0,), ...)
- FakeTensor(size=(s1, s2, s3), ...)
```

In summary, this PR:
- Replaces `SymNode._expr` by `SymNode.expr`, removing the old property function
    - This makes it so `SymNode` instances never update their expression
- Creates `SymNode.simplified_expr()` method for actually calling `ShapeEnv.replace` on
  its expression. Note that this doesn't updates `SymNode.expr`
- Changes how `tensor.size()` gets converted to its Python `torch.Size` type
    - Instead of calling `SymInt::maybe_as_int()` method, we create a new
      `SymInt::is_symbolic()` method for checking whether it is actually a symbolic value
    - This is needed so that when we call `tensor.size()` in the Python side, the returned
      sequence is faithful to the actual data, instead of possibly simplifying it and
      returning an integer
    - 2 files needs this modification:
        - _torch/csrc/Size.cpp_: for handling `torch.Tensor.size` Python calls
        - _torch/csrc/utils/pybind.cpp_: for handling `symint.cast()` C++ calls

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

### CODE REVIEW

File: torch/csrc/Size.cpp [hunk 1]
```diff
@@ -160,8 +165,8 @@ static PyObject* wrap_tuple_fn(Args... args) {
 // We use an anonymous namespace instead of static to work around
 // (what @peterjc123 think is) a bug in Visual Studio
 namespace {
-auto sq_concat = PyTuple_Type.tp_as_sequence -> sq_concat;
-auto sq_repeat = PyTuple_Type.tp_as_sequence -> sq_repeat;
+auto sq_concat = PyTuple_Type.tp_as_sequence->sq_concat;
+auto sq_repeat = PyTuple_Type.tp_as_sequence->sq_repeat;
```
Comment (ysiraichi) [inline torch/csrc/Size.cpp:169]: Linter did this.

File: torch/csrc/Size.cpp [hunk 2]
```diff
@@ -55,7 +55,20 @@ PyObject* THPSize_NewFromSymSizes(const at::Tensor& self_) {
 
   for (auto i : c10::irange(sym_sizes.size())) {
     auto si = sym_sizes[i];
-    if (auto m = si.maybe_as_int()) {
+    if (si.is_symbolic()) {
```
Comment (ysiraichi) [inline torch/csrc/Size.cpp:58]: I still have to the same thing for stride and storage_offset.

File: c10/core/SymInt.h [hunk 3]
```diff
@@ -145,6 +145,12 @@ class C10_API SymInt {
   // number can be used to diagnose overspecialization.
   int64_t guard_int(const char* file, int64_t line) const;
 
+  // Distinguish actual symbolic values from large negative integers.
+  bool is_symbolic() const {
+    return is_heap_allocated() &&
+        toSymNodeImplUnowned()->large_negative_int() == 0;
```
Comment (ezyang) [inline c10/core/SymInt.h:151]: @soulitzer This will conflict with your stuff

File: test/test_dynamic_shapes.py [hunk 4]
```diff
@@ -195,13 +195,17 @@ def test_roundtrip(self):
 
         self.assertTrue(x.size()[0], 5)
         self.assertTrue(x.size()[1], 4)
-        self.assertTrue(isinstance(x.size()[1], int))  # due to guard above
+        # Should be simplifiable to an integer.
+        # Ref: https://github.com/pytorch/pytorch/pull/107492
+        self.assertTrue(isinstance(x.size()[1], SymInt))
```
Comment (ezyang) [inline test/test_dynamic_shapes.py:200]: This seems worse; we specifically wanted specialized SymInt to decay to int as soon as possible?

**Review by ezyang — COMMENTED** [review]
Can you benchmark the change?

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -2654,8 +2647,10 @@ def is_dim(src):
         # pruned).  In this case, only Dynamo knows about these arguments.
         def track_symint(source, val, constraint=None):
             if isinstance(val, SymInt):
-                s = val.node.expr
+                val = val.node.maybe_as_int() or val
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:2650]: This will wrongly put 0 into val

**Comment by ezyang** [comment]
I triggered a benchmark run here: https://github.com/pytorch/pytorch/actions/runs/5921469587

I see why my version failed and why your version is likely to succeed: you fixed the Python bindings to avoid simplifying early. I'm worried about the performance implications of this. By simplifying early, we get to avoid churning symbolic integers around and skip to fast regular int math. If compile time perf is good we can land this, I can see this also simplifying some of our debug logic too.

**Comment by ysiraichi** [comment]
As far as I can see, it should be about the same, performance-wise. Here's a list of comparisons against commits around the base of this PR.

1. [Comparison against `93f2a64d4d1af38904cb4d36e3509856c205615b` (Aug 19)](https://hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2014%20Aug%202023%2012%3A04%3A31%20GMT&stopTime=Mon%2C%2021%20Aug%202023%2012%3A04%3A31%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/ysiraichi/73/head&lCommit=c6e134d5cdf0a570b04303da6352c9a81f52c2cb&rBranch=main&rCommit=93f2a64d4d1af38904cb4d36e3509856c205615b)
2. [Comparison against `a815e719e85899d4229616617e7827d4de191c2d` (Aug 18)](https://hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2014%20Aug%202023%2012%3A04%3A31%20GMT&stopTime=Mon%2C%2021%20Aug%202023%2012%3A04%3A31%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/ysiraichi/73/head&lCommit=c6e134d5cdf0a570b04303da6352c9a81f52c2cb&rBranch=main&rCommit=a815e719e85899d4229616617e7827d4de191c2d)
3. Base commit: [`36141de42799912538a9c30c38b1a506eb955eed`](https://github.com/pytorch/pytorch/commit/36141de42799912538a9c30c38b1a506eb955eed) (Aug 18)
4. [Comparison against `384e0d104fd077d31efafc564129660e9b7a0f25` (Aug 17)](https://hud.pytorch.org/benchmark/compilers?startTime=Mon%2C%2014%20Aug%202023%2012%3A04%3A31%20GMT&stopTime=Mon%2C%2021%20Aug%202023%2012%3A04%3A31%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/ysiraichi/73/head&lCommit=c6e134d5cdf0a570b04303da6352c9a81f52c2cb&rBranch=main&rCommit=384e0d104fd077d31efafc564129660e9b7a0f25)

File: torch/fx/experimental/symbolic_shapes.py (hunk 5 above)
Comment (ysiraichi) [reply torch/fx/experimental/symbolic_shapes.py:2650]: Not sure I get it. The only problem I can see is if `val.node.maybe_as_int()` returns 0, e.g. if there's a guard `x == 0` (is this possible?).

The way I see it, there are 2 possibilities:
1. `val.node.maybe_as_int()` returns `None`, then `val` will stay as is
2. `val.node.maybe_as_int()` returns an int (other than 0), then `val` will be an int

Am I missing something?

**Review by ezyang — APPROVED** [review]

**Comment by ysiraichi** [comment]
The CI failure seems to be coming from upstream.

**Pull request closed without merging**